tests/controller/controller_test.go

```diff
@@ -0,0 +1,628 @@
+package controller_test
+
+import (
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/awslabs/aws-s3-csi-driver/pkg/podmounter/mppod"
+)
+
+var _ = Describe("Mountpoint Controller", func() {
+	Context("Static Provisioning", func() {
+		Context("Scheduled Pod with pre-bound PV and PVC", func() {
+			It("should schedule a Mountpoint Pod", func() {
+				vol := createVolume()
+				vol.bind()
+
+				pod := createPod(withPVC(vol.pvc))
+				pod.schedule("test-node")
+
+				waitAndVerifyMountpointPodFor(pod, vol)
+			})
+
+			It("should schedule a Mountpoint Pod per PV", func() {
+				vol1 := createVolume()
+				vol1.bind()
+				vol2 := createVolume()
+				vol2.bind()
+
+				pod := createPod(withPVC(vol1.pvc), withPVC(vol2.pvc))
+				pod.schedule("test-node")
+
+				waitAndVerifyMountpointPodFor(pod, vol1)
+				waitAndVerifyMountpointPodFor(pod, vol2)
+			})
+
+			It("should not schedule a Mountpoint Pod if the volume is backed by a different CSI driver", func() {
+				vol := createVolume(withCSIDriver(ebsCSIDriver))
+				vol.bind()
+
+				pod := createPod(withPVC(vol.pvc))
+				pod.schedule("test-node")
+
+				expectNoMountpointPodFor(pod, vol)
+			})
+
+			It("should only schedule Mountpoint Pods for volumes backed by S3 CSI Driver", func() {
+				vol1 := createVolume()
+				vol1.bind()
+				vol2 := createVolume(withCSIDriver(ebsCSIDriver))
+				vol2.bind()
+
+				pod := createPod(withPVC(vol1.pvc), withPVC(vol2.pvc))
+				pod.schedule("test-node")
+
+				waitAndVerifyMountpointPodFor(pod, vol1)
+				expectNoMountpointPodFor(pod, vol2)
+			})
+		})
+
+		Context("Scheduled Pod with late PV and PVC binding", func() {
+			It("should schedule a Mountpoint Pod", func() {
+				vol := createVolume()
+
+				pod := createPod(withPVC(vol.pvc))
+				pod.schedule("test-node")
+
+				expectNoMountpointPodFor(pod, vol)
+
+				vol.bind()
```

Comment:
Semantically, what does this mean in terms of k8s runtime? Where the binding isn't explicit in the PV, but gets bound at pod schedule time?

Reply:
This is actually the default behavior if you don't pre-bind your PVC and PV. You specify a PVC in your Pod with some capacity, and once your Pod scheduled, Kubernetes tries to bind a PV for your PVC that fulfills your capacity requests. In case of static provisioning, cluster admin must pre-provision PVs for this to work, in case of dynamic provisioning another controller like [external-provisioner](https://github.com/kubernetes-csi/external-provisioner) sees that and dynamically creates a PV and binds that to that PVC. See for more detail https://kubernetes.io/docs/concepts/storage/persistent-volumes/#binding